Initial Auditing Configuration and Usage of the kubernetes.io GCP Organization

hh opened the pull request:
```feature
As an community security auditor
I want to retrieve the iam policies and permissions for the kubernetes.io GCP organisation
In order to  provide transparency into who can access and configure community maintained resources

Given an gcloud account within the k8s-infra-gcp-auditors google group
When ./audit.sh is run and a PR created with the resulting policy changes
Then iam configuration can be peered reviewed / audited
```

hh commented:
Initial audit of a policy change: https://kubernetes.slack.com/messages/CCK68P2Q2/convo/C09QZ4DQB-1554150053.006600/

thockin commented on audit/audit.sh:
```diff
@@ -0,0 +1,43 @@
+do
+    gcloud organizations get-iam-policy $CNCF_GCP_ORG --format=$format \
+           > cncf-org-policy.$format
+    gcloud projects get-iam-policy kubernetes-public --format=$format \
```
Rather than "kubernetes-public" shouldn't we be doing a loop over all projects?

e.g.

```
gcloud projects list \
    --filter "parent.id=758905017065" \
    --format "value(name, projectNumber)" \
    | while read NAME NUM; do \
        gcloud projects get-iam-policy $NAME --format=$format > $NAME.policy.$format
        gcloud iam roles list --project=$NAME --format=$format > $NAME.roles.$format
    done
````

thockin replied:
Also, I vote for JSON - much easier to visually verify

hh replied:
Currently doing json and yaml, I can remove yaml if you like.

thockin commented on audit/audit.sh:
```diff
@@ -0,0 +1,43 @@
+for format in json yaml
+do
+    gcloud organizations get-iam-policy $CNCF_GCP_ORG --format=$format \
+           > cncf-org-policy.$format
```
nit: can we use file names with periods as delimiters, since project names can't include periods?  E.g. `cncf-org.policy.$format`  so `*.policy.*` is all policy lists and `*.roles.*` is all roles lists

thockin commented on audit/audit.sh:
```diff
@@ -0,0 +1,43 @@
+
+# Permissions per project role
+mkdir -p roles
+for ROLE_PATH in `gcloud --project=kubernetes-public iam roles list --format="value(name)"`
```
same loop for projects

thockin commented on audit/audit.sh:
```diff
@@ -0,0 +1,43 @@
+mkdir -p buckets
+for BUCKET in `gsutil ls -p kubernetes-public | awk -F/ '{print $3}'`
+do
+    gsutil ls -r gs://$BUCKET/ > buckets/$BUCKET.txt
```
Do we need to audit the contents of buckets, or just permissions?  This will trigger a lot of updates -- e.g. billing is dumped at least daily.

hh replied:
Just permissions are fine, I think there are object/file/folder specific permissions that can be set.
We may need a way to capture these or set a remove the ability to set per file perms and only have bucket level.
Thoughts?

thockin commented on audit/audit.sh:
```diff
@@ -0,0 +1,43 @@
+
+# List of objets in buckets
+mkdir -p buckets
+for BUCKET in `gsutil ls -p kubernetes-public | awk -F/ '{print $3}'`
```
same loop for projects

thockin commented:
Ping on this.  As we start expanding scope, the fully formed version of this would make me feel better...

spiffxp commented:
@thockin to take a final pass on this

hh commented on audit/README.md:
```diff
@@ -0,0 +1,55 @@
+
+We mostly host it in Google Cloud:
+  * GCP org = kubernetes.io / organizationId=758905017065
+  * GCP project = kubernetes-public
```
List removed from readme

hh commented on audit/README.md:
```diff
@@ -0,0 +1,55 @@
+
+### audit.sh
+
+Run ./audit.sh to generate a current audit configuration dump.
```
Done.

hh commented on audit/README.md:
```diff
@@ -0,0 +1,55 @@
+First, the requsting auditor opens a PR with any updates applied to the appropriate YAML/JSON file.
+Next, the requesting auditor validates that the PR looks correct for their request and responds `/lgtm`
+
+The a secondary auditor merges the PR once it has been LGTM'd
```
It is in the Performing an Audit section and the top of README.

hh commented on audit/audit-gcp.sh:
```diff
@@ -0,0 +1,44 @@
+
+# gcloud organizations describe $CNCF_GCP_ORG 2>&1
+# ERROR: (gcloud.organizations.describe)
+# User [hh@ii.coop] does not have permission to access organization []
```
@thockin Can we add 'gcloud.organizations.describe' to the k8s-infra-gcp-auditors?

thockin replied:
That's not the permission name.  I added 'Organization Viewer' which seems to have the right permissions.  Try it and let me know?

thockin replied:
did you try this again?

hh commented on audit/k8s-infra-dev-cluster-turnup.services.yaml:
```diff
@@ -0,0 +1,358 @@
+---
```
Note that the dev-cluster-turnup has a many more services enabled than our primary one.

hh commented on audit/audit-gcp.sh:
```diff
@@ -0,0 +1,92 @@
+    do
+        case $service in
+            compute.googleapis.com)
+                echo TODO: Needs compute.projects.get
```
'compute.projects.get' will also be necessary for 'gcloud compute project-info describe'

thockin replied:
Added compute viewer

hh commented on audit/audit-gcp.sh:
```diff
@@ -0,0 +1,92 @@
+                gcloud dns managed-zones list --format=$format > dns/$PROJECT.zones.$format
+                ;;
+            logging.googleapis.com)
+                echo TODO: Needs serviceusage.services.use
```
serviceusage.services.use is needed for inspecting the usage various services, including logging.

thockin replied:
Added service usage consumer

hh commented:
I dropped yaml, so we only have to look at json.
I also added initial iteration over services, but we'll need a few more permissions.
I don't think the service iterations are a must have for the initial release.

hh commented on audit/audit-gcp.sh:
```diff
@@ -0,0 +1,92 @@
+                echo TODO: Verify how Big Query is configured / audited
+                ;;
+            storage-api.googleapis.com)
+                echo TODO: Add storage.buckets.get for auditors
```
Need storage.buckets.get to ensure bucket-policy-only

thockin replied:
Added

hh commented on audit/audit-gcp.sh:
```diff
@@ -0,0 +1,92 @@
+            storage-api.googleapis.com)
+                echo TODO: Add storage.buckets.get for auditors
+                echo ...to kubernetes_public_billing and any newer buckets...
+                echo TODO: Ensure bucket-policy-only, for simplicity in Auditing
```
@thockin @spiffxp Would likely be good to configure bucket-policy-only for all our buckets. I can't think of a reason we'd need to have object specific perms for now. We could enable it as needed later, but it would simplify auditing for now.

thockin replied:
We should have that on already except GCR-fronted buckets.  If you are seeing that not true, tell me whch

hh replied:
These are the commands that would help simplify auditing... if we disable per object IAM and enforce bucket only:

```
# Check that each bucket only allows per bucket policies
gsutil bucketpolicyonly get gs://$BUCKET/                                                                  
# Configuration / Ensure the content can be accessed from anywhere
gsutil cors get gs://$BUCKET/     
# Ensure and understand where the logging for each bucket is going                                                                         
gsutil logging get gs://$BUCKET/  
```

thockin replied:
We do set `bucketpolicyonly` now.  We do not set logging and we do not set cors.  Should we?

hh commented on audit/audit-gcp.sh:
```diff
@@ -0,0 +1,100 @@
+    for service in `gcloud services list --filter state:ENABLED --format=json \
+                    | jq -r .[].config.name | sed s:.googleapis.com::`
+    do
+        case $service in
```
We do a bit of iterating over the services. How far do we want to dig into the configuration of each.

thockin replied:
I am less concerned about service configs than IAM and access.

hh replied:
Understood. Dumping the entire config is a bit out of scope, but I added it the documentation/TODO completeness later.

spiffxp commented:
/cc

spiffxp commented:
this has grown quite a lot, do we need to open up a new PR?

spiffxp commented:
@thockin @hh WDYT about merging this as-is and iterating from there?

hh commented:
For each TODO we have the permissions required and the commands that require those permissions.

hh commented:
Since adding a few new projects the case statement iterating over services has more unhandled services.
```
cat log.mkd | grep Unhandled | sort | uniq
# Unhandled Service admin #
# Unhandled Service bigtable #
# Unhandled Service bigtableadmin #
# Unhandled Service cloudapis #
# Unhandled Service cloudbuild #
# Unhandled Service clouddebugger #
# Unhandled Service cloudresourcemanager #
# Unhandled Service cloudscheduler #
# Unhandled Service cloudtrace #
# Unhandled Service container #
# Unhandled Service containeranalysis #
# Unhandled Service containerregistry #
# Unhandled Service containerscanning #
# Unhandled Service datastore #
# Unhandled Service groupssettings #
# Unhandled Service iam #
# Unhandled Service iamcredentials #
# Unhandled Service pubsub #
# Unhandled Service servicemanagement #
# Unhandled Service serviceusage #
# Unhandled Service sql-component #
# Unhandled Service stackdriver #
```

hh commented:
These are the new perms and the commands are commented out in audit-gcp under each TODO:
```
TODO: compute Needs compute.projects.get
TODO: Ensure bucket-policy-only, for simplicity in Auditing
TODO: logging needs serviceusage.services.use
TODO: monitoring needs serviceusage.services.use
TODO: storage-api needs storage.buckets.get for auditors
TODO: Verify how Big Query is configured / audited
TODO: Verify how OS Login is configured / audited
```

hh commented:
I'm keen with merging as is and evolving.

spiffxp commented:
/approve
/lgtm
/hold
I'm going to leave the `/hold cancel` to @thockin

thockin commented on audit/audit-gcp.sh:
```diff
@@ -0,0 +1,100 @@
+#!/bin/bash
+# set -x -e
+CNCF_GCP_ORG=758905017065
```
You could source `infra/gcp/lib.sh` if there are any more utilities to re-use.  This is unlikely to change :)

thockin commented on audit/audit-gcp.sh:
```diff
@@ -0,0 +1,100 @@
+#!/bin/bash
```
need copyright header

thockin commented on audit/audit-gcp.sh:
```diff
@@ -0,0 +1,100 @@
+    | while read PROJECT NUM; do \
+    export CLOUDSDK_CORE_PROJECT=$PROJECT
+    echo "### Auditing Project: ${PROJECT} ###"
+    mkdir -p $PROJECT
```
Since this assumes CWD is the script's root, maybe we should check that and print a warning if it is not?

thockin commented on audit/buckets/kubernetes-public.kubernetes_public_billing.txt:
```diff
@@ -0,0 +1,156 @@
+gs://kubernetes_public_billing/billing--2019-01-10.csv
```
We probably should not be dumping bucket contents or VM names or anything that is likely to be constantly changing.  Unlikely that any 2 audits could produce the same data for this.

thockin commented on audit/audit-gcp.sh:
```diff
@@ -0,0 +1,100 @@
+                    #### gsutil cors get gs://$BUCKET/
+                    #### gsutil logging get gs://$BUCKET/
+                    gsutil iam get gs://$BUCKET/ > $PROJECT/buckets/$BUCKET.iam.json
+                    gsutil ls -r gs://$BUCKET/ > $PROJECT/buckets/$BUCKET.txt
```
We should not do this - it will be always changing

thockin commented on audit/audit-gcp.sh:
```diff
@@ -0,0 +1,100 @@
+            compute)
+                echo TODO: $service Needs compute.projects.get
+                #### gcloud compute project-info describe
+                #### gcloud compute instances list --format=$format > $PROJECT/services/compute.instances.$format
```
we should not list VMs - they will change as clusters get upgraded

thockin replied:
disks are somewhat dubiously useful, too.  Clusters seems OK

thockin commented on audit/audit-gcp.sh:
```diff
@@ -0,0 +1,100 @@
+                echo Processing: $service
+                mkdir -p dns
+                gcloud dns project-info describe $PROJECT --format=$format > $PROJECT/services/dns.info.$format
+                gcloud dns managed-zones list --format=$format > $PROJECT/services/dns.zones.$format
```
same as above - audit should catch metadata, not data

thockin commented on audit/k8s-artifacts-graveyard/buckets/asia.artifacts.k8s-artifacts-graveyard.appspot.com.iam.json:
```diff
@@ -0,0 +1,33 @@
+{
+  "bindings": [
```
graveyard is going away, so we can nuke all these files

dims commented:
/assign @hh

spiffxp commented:
merge and iterate

thockin commented:
Ping - I think it's back in your court, @hh

hh commented:
/hold

hh commented:
/hold cancel

hh commented:
/test tide

hh commented:
/retest
